### PR TITLE
Report FlowAnalysis errors in script submissions

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by tannergooding** [comment]
test win please

File: src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs [hunk 1]
```diff
@@ -23,18 +24,41 @@ internal class FlowAnalysisPass
...
             DiagnosticBag diagnostics)
         {
+            if (method.IsScriptInitializer)
+            {
+                // For script content, we're interested in errors but
+                // not warnings (such as unused variable) that are
+                // common when compiling code snippets.
+                var allDiagnostics = DiagnosticBag.GetInstance();
+                block = Rewrite(method, block, addImplicitReturnOnly: false, diagnostics: allDiagnostics);
+                diagnostics.AddRange(allDiagnostics.AsEnumerable().Where(d => d.Severity == DiagnosticSeverity.Error));
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs]: Please check with the diagnostics team. It doesn't seem right to hard-code a blanket policy in such a low layer.

**Comment by cston** [comment]
@tmat, @gafter, @amcasey, @kevinh-ms, @genlu, @ManishJayaswal

File: src/Compilers/CSharp/Test/Semantic/Semantics/ScriptSemanticsTests.cs [hunk 2]
```diff
@@ -712,24 +712,38 @@ public void ERR_VariableUsedBeforeDeclaration()
...
+            c.VerifyDiagnostics(
                 // (1,1): error CS7020: You cannot use 'return' in top-level script code
-                Diagnostic(ErrorCode.ERR_ReturnNotAllowedInScript, "return"));
-
-            var c2 = CreateSubmission("return 17;");
+                // return;
+                Diagnostic(ErrorCode.ERR_ReturnNotAllowedInScript, "return").WithLocation(1, 1),
+                // (1,1): warning CS0162: Unreachable code detected
+                // return;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(1, 1));
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ScriptSemanticsTests.cs]: How is it unreachable?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ScriptSemanticsTests.cs [hunk 3]
```diff
@@ -712,24 +712,38 @@ public void ERR_VariableUsedBeforeDeclaration()
...
+        {
+            var c = CreateSubmission("return 17;");
+            c.VerifyDiagnostics(
                 // (1,1): error CS7020: You cannot use 'return' in top-level script code
-                Diagnostic(ErrorCode.ERR_ReturnNotAllowedInScript, "return"));
+                // return 17;
+                Diagnostic(ErrorCode.ERR_ReturnNotAllowedInScript, "return").WithLocation(1, 1),
+                // (1,1): warning CS0162: Unreachable code detected
+                // return 17;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(1, 1));
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ScriptSemanticsTests.cs]: How is it unreachable?

File: src/Scripting/Core/ScriptBuilder.cs [hunk 4]
```diff
@@ -73,20 +73,13 @@ public int GenerateSubmissionId(out string assemblyName, out string typeName)
             {
                 // get compilation diagnostics first.
                 diagnostics.AddRange(compilation.GetParseDiagnostics());
-                if (diagnostics.HasAnyErrors())
-                {
-                    CompilationError(diagnostics, compiler.DiagnosticFormatter);
-                }
-
+                ThrowIfAnyCompilationErrors(diagnostics, compiler.DiagnosticFormatter);
                 diagnostics.Clear();
```
Comment (gafter) [inline src/Scripting/Core/ScriptBuilder.cs]: Hey, where did the warnings go?

File: src/Scripting/Core/ScriptBuilder.cs [hunk 5]
```diff
@@ -73,20 +73,13 @@ public int GenerateSubmissionId(out string assemblyName, out string typeName)
...
 
                 var executor = Build<T>(compilation, diagnostics, cancellationToken);
 
                 // emit can fail due to compilation errors or because there is nothing to emit:
-                if (diagnostics.HasAnyErrors())
-                {
-                    CompilationError(diagnostics, compiler.DiagnosticFormatter);
-                }
+                ThrowIfAnyCompilationErrors(diagnostics, compiler.DiagnosticFormatter);
```
Comment (gafter) [inline src/Scripting/Core/ScriptBuilder.cs]: Hey, where are the warnings reported?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ScriptSemanticsTests.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ScriptSemanticsTests.cs]: That's actually an error for the implicit return added to Script.&lt;Initialize&gt;. Clearly this is not ideal.

Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ScriptSemanticsTests.cs]: It has the location of the last statement.

If the user wrote an infinite loop, they would see this error, then? That's not ideal.

File: src/Scripting/Core/ScriptBuilder.cs (hunk 4 above)
Comment (cston) [inline src/Scripting/Core/ScriptBuilder.cs]: We were only reporting warnings previously if there were errors as well. Is that the desired behavior?

File: src/Compilers/CSharp/Portable/Lowering/InitializerRewriter.cs [hunk 6]
```diff
@@ -67,6 +66,17 @@ internal static BoundTypeOrInstanceInitializers RewriteScriptInitializer(Immutab
             return new BoundTypeOrInstanceInitializers(syntax, boundStatements.ToImmutableAndFree());
         }
 
+        /// <summary>
+        /// Returns the expression if the statement is actually an
+        /// expression (ExpressionStatementSyntax with no trailing semicolon).
+        /// </summary>
+        internal static BoundExpression GetTrailingScriptExpression(this BoundStatement statement)
```
Comment (tmat) [inline src/Compilers/CSharp/Portable/Lowering/InitializerRewriter.cs]: Making it an extension method seems unnecessary.

File: src/Scripting/Core/ScriptBuilder.cs (hunk 4 above)
Comment (gafter) [inline src/Scripting/Core/ScriptBuilder.cs]: I can imagine only reporting warnings if there are *no* errors, but generally speaking warnings should be reported unless the user suppresses them.

Comment (tmat) [inline src/Scripting/Core/ScriptBuilder.cs]: @gafter When you want to run a script warnings should be ignored. If you building a script they should be reported. Let's keep Chuck's change as is, I'm going to follow up with an appropriate scripting API change.

File: src/Compilers/CSharp/Portable/Lowering/InitializerRewriter.cs (hunk 6 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/InitializerRewriter.cs]: Changed.

**Comment by gafter** [comment]
LGTM. It would be nice to see a test that warnings *are* reported when *building* a script.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ScriptSemanticsTests.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ScriptSemanticsTests.cs]: Logged https://github.com/dotnet/roslyn/issues/5297.

**Comment by tmat** [comment]
:+1:

**Comment by cston** [comment]
Fixes #168, #5279

**Pull request merged**